library(qvalue)
library(lfmm)
########################################################################
# 
# File    : LFMMwild.R
# History : 2019/03/20 Created by Kevin Freeman (KF) 
# 
#----------------------------------------------------------------------
# This script is used to perform LFMM ridge analysis  for association 
# between genotype and environmental factors in wild type populations. 
# It starts with  a genotype matrix generated by Bodie with 
# populationStructureScript.R and subsets it to only include wild type
# populations
#########################################################################
setwd("/media/kevin/TOSHIBA_EXT/LOTTERHOS_LAB/OysterGenomeProject/popstructureOutliers")
### Read in RDS and metadata
allData        <- readRDS("data/genotypeMatrix_excluding_LM.rds")
metadata       <- read.csv("data/modified_samplemetadata.csv", stringsAsFactors = FALSE, header = TRUE)
envi_metadata  <- read.csv("data/environment/full_sample_metadata_4_20_19_ER.csv", stringsAsFactors = FALSE, header = TRUE)

# make the Wild.Sel column match between the two metadata files
envi_metadata$Wild.Sel[which(envi_metadata$Wild.Sel == "inbred")] <- "I"

# combine the two metadata csv files
# NG_H0H4 is dropped because it is only in modified_samplemetadata.csv
# LM_2 is dropped because it is only in full_sample_metadata_4_20_19_ER.csv
common_cols <- intersect(names(envi_metadata), names(metadata))
common_cols <- common_cols[! common_cols %in% c("NOTES", "Ecotype", "Sex")]  # remove columns that don't have same info b/w files
comb_metadata <- merge(metadata, envi_metadata, by = common_cols)

### Select only Wild populations to rewrite matrix
wild        <- allData
wild$G      <- allData$G[, which(comb_metadata$Wild.Sel == "W")]
wild$Pop.ID <- metadata$Pop.ID[which(comb_metadata$Wild.Sel == "W")]

# The genotype matrix has some fixed values, remove them before proceeding
variances <- apply(wild$G, 1, var)
not_fixed <- variances > 0
wild$G <- wild$G[not_fixed, ]
#  update the positions and chr
wild$Pos <- wild$Pos[not_fixed]
wild$Chr <- wild$Chr[not_fixed]

### Save the new genotype matrix just in case
saveRDS(wild, paste("data", "genotypeMatrix_selecting_Wild.rds", sep="/"))

# Start with chromosome 1
geno_chr1 <- wild$G[which(wild$Chr == 1), ]
# scale genotype matrix
scaled.genotype <- scale(as.matrix(t(geno_chr1)))

# create a temperature matrix
temp        <- comb_metadata$Mean_Annual_Temperature_Celsius[which(comb_metadata$Wild.Sel == "W")]
temp_matrix <- matrix(data = temp, nrow = length(temp), ncol = 1)
scaled.temp <- scale(temp_matrix)

# build lfmm ridge model
lfmm.ridge <- lfmm::lfmm_ridge(Y = scaled.genotype, X = scaled.temp, K = 3, lambda = 1e-4)
# perform association testing
lfmm.test.ridge <- lfmm::lfmm_test(Y = scaled.genotype, X = scaled.temp, lfmm = lfmm.ridge, calibrate = "gif")

# save the ridge objects, now it makes sense to move from comp5 to my local machine to inspect the results in a GUI
saveRDS(lfmm.ridge, paste("data", "lfmm_ridge_results_mean_annual_temp.rds", sep = "/"))
saveRDS(lfmm.test.ridge, paste("data", "lfmm_ridge_test_results_mean_annual_temp.rds", sep = "/"))

# read in ridge objects
lfmm.ridge      <- readRDS("data/lfmm_ridge_results_mean_annual_temp_chr1.rds")
lfmm.test.ridge <- readRDS("data/lfmm_ridge_test_results_mean_annual_temp_chr1.rds")
# read in genotype matrix/positions/chr
wild <- readRDS("data/genotypeMatrix_selecting_Wild.rds")

# get p values
p.values.ridge <- lfmm.test.ridge$calibrated.pvalue
print(c("lfmm.test.ridge$gif", lfmm.test.ridge$gif))

# plots
hist(p.values.ridge, col = "lightgreen", main="LFMM ridge chr1")
qval <- qvalue::qvalue(p.values.ridge)
plot(qval)


LFMM_ridge_0.0_Chr1_log10p <- -log10(as.numeric(p.values.ridge))
plot(wild$Pos[which(wild$Chr == 1)], LFMM_ridge_0.0_Chr1_log10p, xlab = "Pos", main = "Chr1 Mean Annual Temp Association Test")
