
---
output: html_document
---
# LD Analysis using 'Combined.SNP.TRSdp5g1FnDNAmaf052alleles.vcf.gz'

## Introduction

This document will walk through an analysis of the relationships between
chromosome length, number of SNPs (before and after thinning __using a window of 10K SNPS__), 
number of exons and the average amount of LD when using a base pair window of 
500K at each chromosome.

Some of the files and R objects that will be loaded in this documentation will
need to generated using some of the other scripts that I have developed and data
which can be found in the [github repository](https://github.com/jpuritz/OysterGenomeProject/tree/master/popstructureOutliers).

## Loading data

```{r}
# Chromosomes in the order found in the VCF file "Combined.SNP.TRSdp5g1FnDNAmaf052alleles.vcf.gz"

CHR <- c("NC_035789.1", "NC_035780.1", "NC_035781.1", "NC_035782.1", 
         "NC_035783.1", "NC_035784.1", "NC_035785.1", "NC_035786.1", 
         "NC_035787.1", "NC_035788.1")

# Chromosomes in order with their lengths.

chr.length <- list("NC_035789.1" = 32650045,
                   "NC_035780.1" = 65668440,
                   "NC_035781.1" = 61752955,
                   "NC_035782.1" = 77061148,
                   "NC_035783.1" = 59691872,
                   "NC_035784.1" = 98698416,
                   "NC_035785.1" = 51258098,
                   "NC_035786.1" = 57830854,
                   "NC_035787.1" = 75944018,
                   "NC_035788.1" = 104168038)

# Loading the full genotype matrix which has been filtered for fixed #
# heterozygote sites and the necessary metdata (SNP positions and corresponding #
# chromosome). This can be generated by running populationStructureScript.R with
# "Combined.SNP.TRSdp5g1FnDNAmaf052alleles.vcf.gz"

allData <- readRDS("genotypeMatrix.rds")
str(allData)

# Loading the thinned genotype matrix and necessary metadata (same as above).
# This can be also be generated using the same file and data as above.

thinnedData <- readRDS("thinnedMatrixAndMetaData.rds")
str(thinnedData)

# The following files are not located on github due to size. You can find them
# on the KITT server, generate them yourself by running ldAnalysis.sh using the
# VCF file or on Google Drive. 

# These files contain a calculation of  the squared correlation coefficient #
# between genotypes encoded as 0, 1 and 2 to represent the number of non-reference
# alleles in each individual (which is LD). The first and digit in the file name
# is the minimum and maximum number of physical bases between the SNPs being #
# tested for LD, respectively.

ld.data.50bp <- read.csv("ldAnalysisFiles\\geno_ld_window_50-50.geno.ld", 
                         sep="\t",
                         header=TRUE,
                         stringsAsFactors=FALSE)
ld.data.200.250bp <- read.csv("ldAnalysisFiles\\geno_ld_window_200-250.geno.ld", 
                         sep="\t",
                         header=TRUE,
                         stringsAsFactors=FALSE)                         
ld.data.450.500bp <- read.csv("ldAnalysisFiles\\geno_ld_window_450-500.geno.ld", 
                         sep="\t",
                         header=TRUE,
                         stringsAsFactors=FALSE)                         
ld.data.4500.5000bp <- read.csv("ldAnalysisFiles\\geno_ld_window_4500-5000.geno.ld", 
                         sep="\t",
                         header=TRUE,
                         stringsAsFactors=FALSE)                         
ld.data.9500.10kbp <- read.csv("ldAnalysisFiles\\geno_ld_window_9500-10000.geno.ld", 
                         sep="\t",
                         header=TRUE,
                         stringsAsFactors=FALSE)                         
ld.data.49500.50kbp <- read.csv("ldAnalysisFiles\\geno_ld_window_49500-50000.geno.ld", 
                         sep="\t",
                         header=TRUE,
                         stringsAsFactors=FALSE)                         
ld.data.99500.100kbp <- read.csv("ldAnalysisFiles\\geno_ld_window_99500-100000.geno.ld", 
                                 sep="\t",
                                 header=TRUE,
                                 stringsAsFactors=FALSE)                         
ld.data.499500.500kbp <- read.csv("ldAnalysisFiles\\geno_ld_window_499500-500000.geno.ld", 
                                  sep="\t",
                                  header=TRUE,
                                  stringsAsFactors=FALSE)                         

# Holding the LD calculation files in a list with a name corresponding to the
# maximum window size used.

ld.vars <- list("50" = ld.data.50bp,
                "200" = ld.data.200.250bp, 
                "500" = ld.data.450.500bp, 
                "5000" = ld.data.4500.5000bp,
                "10000" = ld.data.9500.10kbp,
                "50000" = ld.data.49500.50kbp,
                "100000" = ld.data.99500.100kbp,
                "500000" = ld.data.499500.500kbp)
```
## Processing data

Now that we have all the data, I will (1) verify that all chromosomes are
represented in the LD calculation files and get the digit in the chromosome name
that identifies it (e.g. "NC_035789.1" is only different from other chromosomes
by the 9th character), (2) get the mean LD for each chromosome for each window
size into a dataframe and (3) combine all the necessary data into one dataframe
used to create the plots. There will one main plot that will show the LD decay
for all chromosomes as the window size used in the LD calculations is increased
and six sub plots that demonstrate various trends in the data.

### Verifying number and storing index of chromosomes in files

```{r}
# Verifying that all chromosomes are in the LD calculations

chr.num <- length(unique(ld.vars[[1]]$CHR))
for (i in length(ld.vars)) {
    if (length(unique(ld.vars[[i]]$CHR)) != chr.num) {
        stop("The number of chromosomes in the datasets do not match")
    } else {

    }
}

# Getting the unique character that identifies each chromosome

chr.index <- rep(NA, chr.num)
for (i in 1:chr.num){
    chr.index[i] <- substr(unique(ld.vars[[1]]$CHR)[i], start = 9, stop = 9)
}
```

### Processing data for main plot and sub plots

```{r}

# Create an empty dataframe for main plot
ld.stats <- data.frame(matrix(NA, length(ld.vars), chr.num))

# First column will hold the corresponding maximum window size.
ld.stats[,1] <- as.numeric(names(ld.vars))
colnames(ld.stats)[1] <- "windowSize"

# Okay, this is one part that might be a bit confusing but here we go...

# I loop through the LD calculation files...
for (i in 1:length(ld.vars)) {
    # Convert the LD values to numeric becuase they are originally characters
    ld.vars[[i]]$R.2 <- as.numeric(ld.vars[[i]]$R.2)
    #Then, I loop through each chromosome index number...
    for (j in 1:chr.num) {
        # Then, I loop through each file, calculate the mean LD value for each 
        # chromosome and insert that value into the appropriate location of the 
        # empty dataframe I made earlier.
        ld.stats[i, j+1] <- summary(ld.vars[[i]]$R.2[which(ld.vars[[i]]$CHR == paste("NC_03578", chr.index[j], ".1", sep=""))])[4]
        # And give each column the name of the chromosome that the mean LD values belong to...
        colnames(ld.stats)[j+1] <- paste("NC_03578", chr.index[j], ".1", sep="")
    }
}

# Now, onto the code combining all the data we need for the sub plots...

# Create the dataframe beginning with the chromosome lengths
dataForPlots <- data.frame(as.integer(chr.length))
# add some rownames to help keep track of what data goes where...
rownames(dataForPlots) <- c("NC_035789.1", "NC_035780.1", "NC_035781.1", "NC_035782.1", "NC_035783.1", "NC_035784.1", "NC_035785.1", "NC_035786.1", "NC_035787.1", "NC_035788.1")

# Here lie another part that might be a bit confusing (a lot of subsetting). One
# thing that I should mention here is that chromosomes in thinnedData have
# already been converted to factors...

# I am going to loop through the amount of chromosomes that we have (10)...
for (i in 1:10){
    # First, I will get the number of loci for each chromosome that still remain after pruning
    dataForPlots[which(as.integer(factor(rownames(dataForPlots))) == i), 2] <- length(thinnedData$chromosome[which(thinnedData$chromosome == i)])
    # number of loci for each chromosome prior to pruning
    dataForPlots[which(as.integer(factor(rownames(dataForPlots))) == i), 3] <- length(allData$chromosome[which(allData$chromosome == i)])
    # mean LD for each chromosome 
    dataForPlots[which(as.integer(factor(rownames(dataForPlots))) == i), 4] <- ld.stats[which(ld.stats$windowSize == 500000), which(as.integer(factor(colnames(ld.stats))) == i)]
    # Number of loci for each chromosome (after pruning) per length of each chromosome
    dataForPlots[which(as.integer(factor(rownames(dataForPlots))) == i), 5] <- length(thinnedData$chromosome[which(thinnedData$chromosome == i)]) / as.integer(chr.length[which(as.integer(factor(names(chr.length))) == i)])
    # Number of loci for each chromosome (after pruning) per number of loci of each chromosome (prior to pruning).
    dataForPlots[which(as.integer(factor(rownames(dataForPlots))) == i), 6] <- length(thinnedData$chromosome[which(thinnedData$chromosome == i)]) / length(allData$chromosome[which(allData$chromosome == i)])
}

# Number of exons per length of each chromosome in the order of the chromosomes
# in the VCF file (see thread on oyster genome slack channel)

exonPerLength <- c(0.000302358, 0.000533026, 0.000661296, 0.000611488, 
                   0.000611021, 0.000558459, 0.000377209, 0.000405666, 
                   0.000427934, 0.000360494)

dataForPlots[,7] <- exonPerLength

# Add column names
colnames(dataForPlots)  <- c("Length", 
                             "AmountOfSNPsAfterPruning", 
                             "AmountOfSNPsBeforePruning",
                             "LD", 
                             "SnpsPerLength", 
                             "ThinnedSnpsPerOriginalSnps", 
                             "exonsPerLength")
```

```{r}
#Easily distinguishable colors for the plots

col_vector<-c('#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231', 
              '#911eb4', '#46f0f0', '#f032e6', '#bcf60c', '#fabebe', 
              '#008080', '#e6beff', '#9a6324', '#fffac8', '#800000', 
              '#aaffc3', '#808000', '#ffd8b1', '#000075', '#808080', 
              '#ffffff', '#000000')
```
## Main plot: LD Decay across window sizes used

```{r, echo = TRUE}
par(bty="l")
par(mar = c(5, 4, 4, 8) + 0.1)
for (i in 2:ncol(ld.stats)) {
    if (i == 2){
        plot(jitter(ld.stats[,i]) ~ ld.stats[, 1],
             pch = i,
             col = col_vector[i],
             cex = 2,
             xlab="Distance (bp)", 
             ylab="LD (r^2)",
             ylim=c(0,0.24),
             type = "b",
             log = "x",
             main="LD Decay")
    } else {
        points(jitter(ld.stats[,i]) ~ ld.stats[, 1],
               pch = i,
               col = col_vector[i],
               cex = 2,
               type = "b")
    }
    legend("topright",
            inset = c(-0.35,0),
            legend = CHR, 
            pch = 2:(length(CHR)+1),
            xpd = TRUE,
            col = col_vector[2:(length(CHR)+1)])
}
dev.copy(png, "ldDecayPlot.png")
dev.off()
```

## Sub plots...

1. Number of SNPs from each chromosome before pruning against the length of each chromosome  

```{r, echo=TRUE}
par(mar = c(5, 4, 4, 8) + 0.1)
plot(dataForPlots$AmountOfSNPsBeforePruning ~ dataForPlots$Length,
     pch = 2:(length(CHR)+1),
     cex = 2,
     col = col_vector[2:(length(CHR)+1)], 
     ylab = "Number of SNPs from Chromosome Before Pruning", 
     xlab = "Length of Chromosome")

legend("topright",
        inset = c(-0.35,0),
        legend = CHR, 
        pch = 2:(length(CHR)+1),
        xpd = TRUE,
        col = col_vector[2:(length(CHR)+1)])
dev.copy(png, "NoOfOriginalSNPsAndLengthPlot.png")
dev.off()
```
1. Number of SNPs from each chromosome after pruning against the length of each chromosome  

```{r, echo=TRUE}
par(mar = c(5, 4, 4, 8) + 0.1)
plot(dataForPlots$AmountOfSNPsAfterPruning ~ dataForPlots$Length,
     pch = 2:(length(CHR)+1),
     cex = 2,
     col = col_vector[2:(length(CHR)+1)], 
     ylab = "Number of SNPs from Chromosome After Pruning", 
     xlab = "Length of Chromosome")

legend("topright",
        inset = c(-0.35,0),
        legend = CHR, 
        pch = 2:(length(CHR)+1),
        xpd = TRUE,
        col = col_vector[2:(length(CHR)+1)])
dev.copy(png, "NoOfThinnedSNPsAndLengthPlot.png")
dev.off()
```


3. Number of pruned SNPs vs amount of LD at 500K

```{r, echo = TRUE}
par(mar = c(5, 4, 4, 8) + 0.1)
plot(dataForPlots$AmountOfSNPsAfterPruning ~ dataForPlots$LD,
     pch = 2:(length(CHR)+1),
     cex = 2,
     col = col_vector[2:(length(CHR)+1)], 
     ylab = "Number of SNPs from Chromosome After Pruning", 
     xlab = "LD @ 500K")

legend("topright",
        inset = c(-0.35,0),
        legend = CHR, 
        pch = 2:(length(CHR)+1),
        xpd = TRUE,
        col = col_vector[2:(length(CHR)+1)])
dev.copy(png, "NoOfThinnedSNPsVsLD.png")
dev.off()
```

4. Amount of LD @ 500K against the length for each chromosome

```{r, echo = TRUE}
# amount of LD at 500K vs length
par(mar = c(5, 4, 4, 8) + 0.1)
plot(dataForPlots$LD ~ dataForPlots$Length,
     pch = 2:(length(CHR)+1),
     cex = 2,
     col = col_vector[2:(length(CHR)+1)], 
     ylab = "LD @ 500K", 
     xlab = "Chromosome Length")

legend("topright",
        inset = c(-0.35,0),
        legend = CHR, 
        pch = 2:(length(CHR)+1),
        xpd = TRUE,
        col = col_vector[2:(length(CHR)+1)])
dev.copy(png, "LDVsChrLength.png")
dev.off()
```

5. Amount of LD at 500K vs exons/length (see thread on oyster genome slack channel) for each chromosome

```{r, echo  = TRUE}
par(mar = c(5, 4, 4, 8) + 0.1)
plot(dataForPlots$LD ~ dataForPlots$exonsPerLength,
     pch = 2:(length(CHR)+1),
     cex = 2,
     col = col_vector[2:(length(CHR)+1)], 
     ylab = "LD @ 500K", 
     xlab = "Exons per Chromosome Length")

legend("topright",
        inset = c(-0.35,0),
        legend = CHR, 
        pch = 2:(length(CHR)+1),
        xpd = TRUE,
        col = col_vector[2:(length(CHR)+1)])
dev.copy(png, "LDVsExonsPerChrLength.png")
dev.off()
```

6. Number of thinned SNPs per number of original SNPs for each chromosome against LD @ 500K 

```{r, echo  = TRUE}
par(mar = c(5, 4, 4, 8) + 0.1)
plot(dataForPlots$ThinnedSnpsPerOriginalSnps ~ dataForPlots$LD,
     pch = 2:(length(CHR)+1),
     cex = 2,
     col = col_vector[2:(length(CHR)+1)], 
     ylab = "(# SNPs left after pruning)/(# original SNPs) @ chromosome", 
     xlab = "LD @ 500K")

legend("topright",
        inset = c(-0.35,0),
        legend = CHR, 
        pch = 2:(length(CHR)+1),
        xpd = TRUE,
        col = col_vector[2:(length(CHR)+1)])
dev.copy(png, "ThinnedSnpsPerOriginalSnpsVsLD.png")
dev.off()
```