# 2019-03-14 - KBW - Progress Update

## Thinning

__There have been two major changes to this portion of the pipeline__

1. I have refactored the populationStructureScript.R to where __thinning can be done while excluding populations__. I did this in order to remove the LM samples from all analyses including the thinning process.
   * This, however, comes in direct contradiction with some of the code that I have written to do the outlier analysis. Before, I was relying on the outlierAnalysis.R script to remove certain samples.
     * This won't affect anything other than where the user specifies to exclude LM samples and file/folder naming.
     * Maybe there is a better solution to this problem? I haven't thought of one yet.
2. I have altered the code to not only remove fixed heterozygotes but all sites that are fixed to be heterozygous or homozygous for the ancestor or derived alleles. The sites which are removed from the analysis are recorded and output to ```data/thinned_snps/fixed_sites_XXXXXXX_XX.txt```

__Window sizes__
* While thinning SNPs of all populations (excluding LM), __I used a window size of 5000 bp__. This window size was indicated by the LD decay analysis. The LD decay analysis plot can be seen in the ```figures/1LD_analysis/``` folder.
* I will be changing this window for analyses looking at Atlantic Wild populations and Atlantic Selection populations.
  * From the looks of the plots, I would recommend a window of 100K.

## Outlier Analysis

After thinning, I ran the outlier analysis on all populations (excluding LM, of course). I have produced the output of the OutFLANK results for the first chromosome to see the column names and the format for the unique identifiers.

### Results

The manhattan plots can be seen in the folder ```data/large_outputs/outlierAnalysis/outlierAnalysis_XXX/plots/outflank.outlierFinder/results/```

#### Unique Identifiers

These can be changed to something else if this isn't the exact format that we agreed upon but I believe that it was. I say this because we had discussed using something like ```Chr_XX_Pos_XXXXX``` but in the powerpoint presentation shown to the group, it shows what I have in the results now.

```R
'data.frame':   506946 obs. of  16 variables:
 $ LocusName                    : int  1617 1849 18601 18606 18712 18716 18718 18733 18746 18778 ...
 $ OutFLANK_0.2_He              : num  0.341 0.313 0.251 0.242 0.48 ...
 $ OutFLANK_0.2_FST             : num  0.0823 0.0718 0.155 0.1647 0.353 ...
 $ OutFLANK_0.2_T1              : num  0.0142 0.0113 0.0197 0.0203 0.0871 ...
 $ OutFLANK_0.2_T2              : num  0.172 0.158 0.127 0.123 0.247 ...
 $ OutFLANK_0.2_FSTNoCorr       : num  0.175 0.127 0.21 0.219 0.404 ...
 $ OutFLANK_0.2_T1NoCorr        : num  0.0302 0.0201 0.0268 0.027 0.0997 ...
 $ OutFLANK_0.2_T2NoCorr        : num  0.172 0.158 0.127 0.123 0.247 ...
 $ OutFLANK_0.2_meanAlleleFreq  : num  0.782 0.806 0.853 0.859 0.4 ...
 $ OutFLANK_0.2_pvalues         : num  0.6761 0.73292 0.37324 0.31547 0.00437 ...
 $ OutFLANK_0.2_pvaluesRightTail: num  0.33805 0.63354 0.18662 0.15773 0.00218 ...
 $ OutFLANK_0.2_qvalues         : num  0.261 0.354 0.199 0.182 0.016 ...
 $ OutFLANK_0.2_OutlierFlag     : logi  FALSE FALSE FALSE FALSE FALSE FALSE ...
 $ Pos                          : int  1617 1849 18601 18606 18712 18716 18718 18733 18746 18778 ...
 $ Chr                          : int  1 1 1 1 1 1 1 1 1 1 ...
 $ unique                       : chr  "01_00001617" "01_00001849" "01_00018601" "01_00018606" ...
```

## Specific Contrasts

I am running these now and will have results hopefully soon.