---
title: "Eastern Oyster All REsults"
author: "Katie Lotterhos"
date: "5/22/2019"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#install.packages("data.table")
#install.packages("tidyverse")
library(data.table)
library(tidyverse)
```

## Upload data files

```{r cars}

# Dataset for GEA associations
  alldat <- fread("/Users/katie/Google Drive/katie_research/___Eastern Oyster Genome/Population Structure/large_outputs/UnrelatedOutlier_WildForAssocEnviAssoc_MergedData_Lotterhos.txt")

#cpgoedat <- readRDS("/Users/katie/Google Drive/katie_research/___Eastern Oyster Genome/Population Structure/large_outputs/envi_assoc_CpG/CpGOE_full_results.rds")

#cpgoedat <- readRDS("https://drive.google.com/file/d/1zruapRIwHrUllnDXS-1biq8Au8ffvSjA/view?usp=sharing")

# Dataset for CpG Expected/Observed
cpgoedat <- fread("/Users/katie/Google Drive/katie_research/___Eastern Oyster Genome/Population Structure/large_outputs/envi_assoc_CpG/CpGOE_full_results.txt")
head(cpgoedat)

# Dataset for salinity differntial expression (Kevin)
  salDE <- fread("/Users/katie/Desktop/Repos/OysterGenomeProject/popstructureOutliers/data/gene_expression/Salinity.Diff.Expression.csv")
  head(salDE)


# Dataset for temperature differential expression (Kevin)
  tempDE <- fread("/Users/katie/Desktop/Repos/OysterGenomeProject/popstructureOutliers/data/gene_expression/Temperature.Diff.Expression.csv")
  head(tempDE)

# Dataset for dermo differential expression (Dina)
  dermoDE <- fread("/Users/katie/Desktop/Repos/OysterGenomeProject/popstructureOutliers/data/gene_expression/Proestou_dermo_regulation_with_position.csv")
  head(dermoDE)

# Other identifying datasets
    gene_id <- fread("/Users/katie/Desktop/Repos/OysterGenomeProject/popstructureOutliers/data/gene_expression/oyster_identifiers_XM_XP_proteinName.csv")
    gene_pos <- fread("/Users/katie/Desktop/Repos/OysterGenomeProject/popstructureOutliers/data/gene_expression/XP_start_end_Single_copy.tab")

    head(gene_id)
    head(gene_pos)

# merge SalDE with gene info
    names(salDE)[1] <- "XM"
    names(salDE)[which(names(salDE)=="start")] <- "start_old"
    names(salDE)[which(names(salDE)=="end")] <- "end_old"
    salDE2 <- left_join(salDE, gene_id)
    salDE3 <- left_join(salDE2, gene_pos)
    head(salDE3)
    dim(salDE3)
    dim(salDE)
      #check dimenstions are the same
    sum(complete.cases(salDE))
    sum(complete.cases(salDE3))
    sum(!complete.cases(salDE3))
    # looks like there's some missing data, let's take a look at it
    salDE3[!complete.cases(salDE3),]
    # looks like there's a lot of XR's will have to ask Kevin about that
    salDE <- salDE3
      # TO DO: check this with Kevin. Something seems fishy here.

# merge tempDE with gene info
  head(tempDE)
  names(tempDE)[1] <- "XM"
  names(tempDE)[which(names(tempDE)=="start")] <- "start_old"
  names(tempDE)[which(names(tempDE)=="end")] <- "end_old"
  tempDE2 <- left_join(tempDE, gene_id)
  tempDE3 <- left_join(tempDE2, gene_pos)
  head(tempDE3)
  dim(tempDE3)
  dim(tempDE)
    #check dimenstions are the same
  sum(complete.cases(tempDE))
  sum(complete.cases(tempDE3))
  sum(!complete.cases(tempDE3))
    # looks like there's some missing data, let's take a look at it
  tempDE3[!complete.cases(tempDE3),]
    # looks like there's a lot of XR's will have to ask Kevin about that
  tempDE <- tempDE3

```

## Including Plots

You can also embed plots, for example:


```{r}

table(alldat$chrom)

mychrom = "NC_035782.1"
dim(alldat)

dim( alldat %>% filter(chrom == mychrom))

datchr <- alldat %>% filter(chrom == mychrom, !SpuriousOutliers_wild_for_assoc,  !(LFMM_ridge_0.0_log10p_Mean_Annual_Temperature_Celsius_wild_for_assoc > 38.5))
  # this is a quick fix to remove the spurious outliers from the analysis
dim(datchr)

y1_points <- datchr$OutFLANK_0.2_bigsnpr_0.8.2_OutlierFlag_unrelated & datchr$pcadapt_4.0.3_bigsnpr_0.8.2_OutlierFlag_unrelated

dat_points <- datchr %>% filter(y1_points)

dim(dat_points)
#head(dat_points)

#j <- ggplot(datchr, aes(x = chromStart, y=y1))
#j + geom_point() + theme_classic()

#j + geom_area() + theme_classic()

#####################################
### Plot FST colored by P-value ####
#####################################
  p <- ggplot(datchr, aes(x = chromStart, y=OutFLANK_0.2_bigsnpr_0.8.2_FST_unrelated)) +
    geom_bin2d(binwidth = c(10000, 0.01), color="grey", alpha=0.5) +
    theme_classic() +
    geom_point(data=dat_points,
               aes(x=chromStart, y=OutFLANK_0.2_bigsnpr_0.8.2_FST_unrelated),
               colour="darkblue", alpha = 0.3,  size=1) +
  ylim(-1,1) + xlim(c(0, max(datchr$chromEnd)*1.1))
p0 <- ggplot(datchr,
            aes(x = chromStart,
                y = OutFLANK_0.2_bigsnpr_0.8.2_FST_unrelated)) + theme(panel.grid.major = element_blank(),
                                                                       panel.grid.minor = element_blank(),
                                                                       panel.background = element_blank(),
                                                                       axis.line = element_line(colour = "black"))

# misassembled - misassembled chromosomes that need to have rectangles added.
misassembled <- c("NC_035784.1", "NC_035785.1", "NC_035788.1")

# breakpoints - data table with the breakpoint locations. Will be used to show
# where these breakpoints are in the plots.
breakpoints <- read.table("breakpoints_for_plots.txt", header = TRUE)

if (mychrom %in% misassembled) {
    brk.tmp <- breakpoints[which(breakpoints$chr.name == mychrom), ]
    for (j in 1:nrow(brk.tmp)) {
        p0 <- p0 + annotate("rect",
                            xmin = brk.tmp$Low[j],
                            xmax = brk.tmp$High[j],
                            ymin = -Inf,
                            ymax = Inf,
                            fill = "grey",
                            alpha = 0.2)
    }
}

# Plot FST colored by P-value
  p1 <- p0  + geom_bin2d(binwidth = c(10000, 0.01), color="lightblue", alpha=0.5) + theme_classic() +
  geom_point(data=dat_points, aes(x=chromStart, y=OutFLANK_0.2_bigsnpr_0.8.2_FST_unrelated), colour="darkblue", alpha = 0.5,  size=1)
 # p

thresh <- 7
#####################################
### highlight temperature outliers ####
#####################################
  y2_points <- datchr$LFMM_ridge_0.0_log10p_Mean_Annual_Temperature_Celsius_wild_for_assoc > thresh
  #hist(datchr$LFMM_ridge_0.0_log10p_Mean_Annual_Temperature_Celsius_wild_for_assoc,breaks=100)
  dat2_points <- datchr %>% filter(y2_points)  
  # check to make sure spurious outliers removed
  table(datchr$LFMM_ridge_0.0_log10p_Mean_Annual_Temperature_Celsius_wild_for_assoc[datchr$LFMM_ridge_0.0_log10p_Mean_Annual_Temperature_Celsius_wild_for_assoc > 30])
p2 <- p1 + geom_point(data=dat2_points, aes(x=chromStart, y=OutFLANK_0.2_bigsnpr_0.8.2_FST_unrelated), colour="black", alpha = 0.5,  size=3, shape = 6) #upsidedown triangle

  # check to make sure no spurious outliers
  ggplot(data=datchr, aes(x=chromStart,
    y=LFMM_ridge_0.0_log10p_Mean_Annual_Temperature_Celsius_wild_for_assoc)) +
    geom_bin2d(binwidth = c(100000, 0.01), , colour="blue", alpha = 0.7) +  
    geom_point(data=dat2_points, aes(x=chromStart,
               y=LFMM_ridge_0.0_log10p_Mean_Annual_Temperature_Celsius_wild_for_assoc),
               colour="darkblue", alpha = 0.5,  size=1)

    # add temperature outliers to plot
  p2 <- p + geom_point(data=dat2_points, aes(x=chromStart,
                                             y=OutFLANK_0.2_bigsnpr_0.8.2_FST_unrelated),
                       colour="sienna1", alpha = 0.9,  size=3, shape = 6) #upsidedown triangle

#####################################    
# highlight salinity outliers
#####################################
  hist(datchr$LFMM_ridge_0.0_log10p_Mean_Annual_Salinity_ppt_wild_for_assoc, breaks=100)
  y3_points <- datchr$LFMM_ridge_0.0_log10p_Mean_Annual_Salinity_ppt_wild_for_assoc > thresh
  dat3_points <- datchr %>% filter(y3_points)  

  # check to make sure no spurious outliers
  ggplot(data=datchr, aes(x=chromStart,
    y=LFMM_ridge_0.0_log10p_Mean_Annual_Salinity_ppt_wild_for_assoc)) +
    geom_bin2d(binwidth = c(100000, 0.01), , colour="blue", alpha = 0.7) +  
    geom_point(data=dat3_points, aes(x=chromStart,
               y=LFMM_ridge_0.0_log10p_Mean_Annual_Salinity_ppt_wild_for_assoc),
               colour="darkblue", alpha = 0.8,  size=1)

  # add salinity layer to plot
  p3 <- p2 + geom_point(data=dat3_points, aes(x=chromStart,
                                              y=OutFLANK_0.2_bigsnpr_0.8.2_FST_unrelated),
                        colour="springgreen2", alpha = 0.8,  size=5, shape=8)
# star shape

#p3    

#####################################   
# highlight dermo outliers
#####################################
  hist(datchr$LFMM_ridge_0.0_log10p_Dermo_pressure_wild_for_assoc)
  y4_points <- datchr$LFMM_ridge_0.0_log10p_Dermo_pressure_wild_for_assoc > thresh
  dat4_points <- datchr %>% filter(y4_points)  

  # check to make sure no spurious outliers
  ggplot() + geom_bin2d(binwidth = c(100000, 0.01), data=datchr,
                        aes(x=chromStart,
                            y=LFMM_ridge_0.0_log10p_Dermo_pressure_wild_for_assoc),
                        colour="blue", alpha = 0.7) +  
    geom_point(data=dat4_points, aes(x=chromStart,
                                     y=LFMM_ridge_0.0_log10p_Dermo_pressure_wild_for_assoc),
               colour="darkblue", alpha = 0.5,  size=1)

  # add dermo layer to plot
  p4 <- p3 + geom_point(data=dat4_points, aes(x=chromStart,
                                              y=OutFLANK_0.2_bigsnpr_0.8.2_FST_unrelated),
                        colour="turquoise4", alpha = 0.7,  size=5, shape=5)
# diamond shape
  p4    
```

### Add gene expression layers

```{r}
#####################################
# highlight gene expression data - salinity
#####################################
  head(salDE)

  salDE_chr <- salDE %>% filter(chr==mychrom)
  head(salDE_chr)
  #number of significant outliers
  table(salDE_chr$sig_logFC.warm_15_25)

  #after playing around a lot with the graphing, I decided not to show non-DE genes

  # First highlight the response to saltier conditions
  p5 <- p4 + #geom_point(data=salDE_chr, aes(x=start + (start - end)/2,
              #                                 y=-0.2), size=4, alpha=0.01, pch=19) +
        geom_point(data=salDE_chr[which(salDE_chr$sig_logFC.warm_15_25),],
                   aes(x=start + (start - end)/2, y = -0.2,
                       color=logFC.warm_15_25), size=4, alpha=0.5, pch=19)+
    geom_point(data=salDE_chr[which(salDE_chr$sig_logFC.warm_15_25),],
                   aes(x=start + (start - end)/2, y = -0.2), size=4, alpha=0.5, pch=21)+
    geom_text(data=data.frame(1,1),aes(x=max(datchr$chromStart), y=-0.2, label="Saltier", hjust=0),
              size=4) +
    scale_color_gradient2(low="blue", high="red", mid="white", midpoint=0,
                          breaks=seq(-15,15, by=1))

   # Next highlight the response to fresher conditions

  table(salDE_chr$sig_logFC.warm_15_7)

  p6 <- p5 + geom_point(data=salDE_chr[which(salDE_chr$sig_logFC.warm_15_7),],
                        aes(x=start + (start - end)/2,
                                               y=-0.3, color=logFC.warm_15_7),  size=4,
                       alpha=0.5, pch=19)+
    geom_point(data=salDE_chr[which(salDE_chr$sig_logFC.warm_15_7),],
                        aes(x=start + (start - end)/2,
                                               y=-0.3),  size=4,
                       alpha=0.5, pch=21)+
    geom_text(data=data.frame(1,1),aes(x=max(datchr$chromStart),  
                                       y=-0.3, label="Fresher"), size=4, hjust=0)
  #p6

#####################################
# highlight gene expression data - temperature
#####################################   
  tempDE_chr <- tempDE %>% filter(chr==mychrom)
  head(tempDE)

#### highlight response to warming in LA Lousianana population
  table(tempDE_chr$sig_LA3010)

  p7 <- p6 + geom_point(data=tempDE_chr[which(tempDE_chr$sig_LA3010),], aes(x=start + (start - end)/2,
                                               y=-0.4, color=logFC.LA_30_10),  size=4, alpha=0.5, pch=19)+
    geom_point(data=tempDE_chr[which(tempDE_chr$sig_LA3010),], aes(x=start + (start - end)/2,
                                               y=-0.4),  size=4, alpha=0.5, pch=21)+
    geom_text(data=data.frame(1,1),aes(x=max(datchr$chromStart),  y=-0.4, label="Warmer-LA"), size=4, adj=0)

#### highlight response to warming in NB New Brunswick population
  table(tempDE_chr$sig_NB3010)

  p8 <- p7 + geom_point(data=tempDE_chr[which(tempDE_chr$sig_NB3010),], aes(x=start + (start - end)/2,
                                               y=-0.5, color=logFC.NB_30_10),  size=4, alpha=0.5, pch=19)+
     geom_point(data=tempDE_chr[which(tempDE_chr$sig_NB3010),], aes(x=start + (start - end)/2,
                                               y=-0.5),  size=4, alpha=0.5, pch=21) +
    geom_text(data=data.frame(1,1),aes(x=max(datchr$chromStart),  y=-0.5, label="Warmer-NB"), size=4, adj=0)


#####################################
# highlight gene expression data - dermo Resistant Line
#####################################  
  dermoDE_chr <- dermoDE %>% filter(chr==mychrom)

  table(dermoDE_chr$res7d_sig)
  hist(dermoDE_chr$res7d_log2FoldChange)
  hist(scale(dermoDE_chr$res7d_log2FoldChange))
  hist(dermoDE_chr$textDermoRes)
  p9 <- p8 + geom_point(data=dermoDE_chr[which(dermoDE_chr$res7d_sig),],
                        aes(x=start + (start - end)/2,
                                               y=-0.6, color=res7d_log2FoldChange),  size=4,  pch=19)+
    geom_point(data=dermoDE_chr[which(dermoDE_chr$res7d_sig),],
                        aes(x=start + (start - end)/2,
                                               y=-0.6),  size=4,  pch=21)+
    geom_text(data=data.frame(1,1),aes(x=max(datchr$chromStart),  y=-0.6, label="Dermo-Resistant"), size=4, adj=0)  
  #p9

#####################################
# highlight gene expression data - dermo susceptible Line
#####################################      

  table(dermoDE_chr$sus7d_sig)
  hist(dermoDE_chr$sus7d_log2FoldChange)
  hist(scale(dermoDE_chr$sus7d_log2FoldChange))

  p10 <- p9 +
    geom_point(data=dermoDE_chr[which(dermoDE_chr$sus7d_sig),], aes(x=start + (start - end)/2,
                                               y=-0.7, color=sus7d_log2FoldChange),  size=4,  pch=19)+
    geom_point(data=dermoDE_chr[which(dermoDE_chr$sus7d_sig),], aes(x=start + (start - end)/2,
                                               y=-0.7),  size=4,  pch=21)+
    geom_text(data=data.frame(1,1),aes(x=max(datchr$chromStart),  y=-0.7, label="Dermo-Susceptible"), size=4, adj=0)  

  #p10

 #####################################
# highlight gene expression data - dermo tolerant Line
#####################################     
  table(dermoDE_chr$tol7d_sig)
  hist(dermoDE_chr$tol7d_log2FoldChange)

  p11 <- p10 +
    geom_point(data=dermoDE_chr[which(dermoDE_chr$tol7d_sig),], aes(x=start + (start - end)/2,
                                               y=-0.8, color=tol7d_log2FoldChange),  size=4,  pch=19)+
    geom_point(data=dermoDE_chr[which(dermoDE_chr$tol7d_sig),], aes(x=start + (start - end)/2,
                                               y=-0.8),  size=4,  pch=21)+
    geom_text(data=data.frame(1,1),aes(x=max(datchr$chromStart),  y=-0.8, label="Dermo-Tolerant"), size=4, adj=0)  

  p11
  # TO DO: ADD LEGEND
# TO DO: ADD LAYER FOR misassemblies
# TO DO: get more data together!
# TO DO: write a script to visualize allele frequncy as function of population or environment (with population symbols)
#
```


#



# highlight temp CpG OE outliers
hist(cpgoedat$LFMM_ridge_0.0_log10p_cpgoe_Mean_Annual_Temperature_Celsius_wild_for_assoc)
y5_points <- cpgoedat$LFMM_ridge_0.0_log10p_cpgoe_Mean_Annual_Temperature_Celsius_wild_for_assoc > 5
dat5_points <- cpgoedat %>% filter(y5_points & seqid==mychrom)  
dim(dat5_points)

p5 <- p4 + geom_point(data=dat5_points, aes(x=start+(start-end)/2, y=FST_meanNoNA), colour="red", alpha = 0.7,  size=3, shape = 6)
#p5

# highlight dermo CpG OE outliers
hist(cpgoedat$LFMM_ridge_0.0_log10p_cpgoe_Dermo_pressure_wild_for_assoc)
y6_points <- cpgoedat$LFMM_ridge_0.0_log10p_cpgoe_Dermo_pressure_wild_for_assoc > 5
dat6_points <- cpgoedat %>% filter(y6_points & seqid==mychrom)  
dim(dat6_points)
p6 <- p5 + geom_point(data=dat6_points, aes(x=start+(start-end)/2, y=FST_meanNoNA), colour="red", alpha = 0.7,  size=5, shape = 5)
#p6

# highlight salinity CpG OE outliers
hist(cpgoedat$LFMM_ridge_0.0_log10p_cpgoe_Mean_Annual_Salinity_ppt_wild_for_assoc)
y7_points <- cpgoedat$LFMM_ridge_0.0_log10p_cpgoe_Dermo_pressure_wild_for_assoc > 3
dat7_points <- cpgoedat %>% filter(y7_points & seqid==mychrom)  
dim(dat7_points)
p7 <- p6 + geom_point(data=dat7_points, aes(x=start+(start-end)/2, y=FST_meanNoNA), colour="red", alpha = 0.7,  size=5, shape = 8)
p7 #+ xlim(3.9e07, 4.2e07)



add.breakpoints <- function (chr.index, ybottom, ytop, ytext) {

    breakpoints <- read.table("breakpoints_for_plots.txt")

    breakpoints <- breakpoints[which(breakpoints$Chr == chr.index), ]

    for (i in 1:nrow(breakpoints)) {
        rect(xleft=breakpoints$Low[i],
             ybottom = ybottom,
             xright = breakpoints$High[i],
             ytop = ytop,
             col="lightgrey",
             border="transparent")
        x_text <- floor((breakpoints$High[i]+breakpoints$Low[i])/2)
        text(x = x_text, y = ytext , paste("LG", breakpoints$LG[i], sep=" "))
    }
}

make.plots <- function (data = NULL, which.vars = NULL, chr.index = NULL, outpath = NA) {

    #######################
    # Load data and Setup #
    #######################

    # data - subset of full data given plots are made chromosome at a time    
    data <- data[which(data$Chr == chr.index), ]

    # plots.path - path for plots to be saved given the subset specified
    plots.path <- outpath

    # data[, negativeLog10p][data[, negativeLog10p] >= 100] <- 100
    # plots.x <- data$Pos[data[, He] > 0.1]
    #
     png(paste(plots.path,
              "/manhattan_plots_",
              sprintf(paste("%0", 2, "d", sep=""),
              chr.index),
              ".png",
              sep = ""),
        height = 1440,
        width = 2048)
    par(mfrow = c(2, 1), mar = c(8, 4, 2, 1))
    plot(plots.x,
         data[, FST][data[, He] > 0.1],
         ylab = "FST",
         xlab = "",
         ylim = c(-0.2, 1.1),
         xlim = c(min(plots.x, na.rm = TRUE), max(plots.x, na.rm = TRUE)),
         main = paste("OutFLANK Manhattan Plot: Chromsome", chr.index),
         col = rgb(0, 0, 0, 0.2),
         xaxt = "n")
    axis(side = 1,
         las = 2,
         at = seq(min(plots.x, na.rm = TRUE),
                  max(plots.x, na.rm = TRUE),
                  length.out = 50))
    mtext(text = "Position (BP)",
          side = 1,
          line = 6)
    # Add misassembly if chromosome has one
    if (chr.index %in% misassembled) {
        add.breakpoints(chr.index, ybottom = -0.25, ytop = 1.15, ytext = 1.1)
        points(data$Pos[data[ , He] > 0.1], data[, FST][data[ , He] > 0.1])
    }
    # Colored points for outliers
    points(data$Pos[which(data[, outflank.outlier] == TRUE)],
           data[, FST][which(data[, outflank.outlier] == TRUE)],
           col = "red",
           pch = 20)

        points(data$Pos[which(data[, outflank.outlier] == TRUE & data[, pcadapt.outlier] != TRUE)],
           data[, FST][which(data[, outflank.outlier] == TRUE & data[, pcadapt.outlier] != TRUE)],
           col = "red",
           pch = 20)
    points(data$Pos[which(data[, pcadapt.outlier] == TRUE & data[, outflank.outlier] != TRUE)],
           data[, FST][which(data[, pcadapt.outlier] == TRUE & data[, outflank.outlier] != TRUE)],
           col = "blue",
           pch = 20)
    points(data$Pos[which(data[, pcadapt.outlier] == TRUE & data[, outflank.outlier] == TRUE)],
           data[, FST][which(data[, pcadapt.outlier] == TRUE & data[, outflank.outlier] == TRUE)],
           col = "orange",
           pch = 20)`


}



# Plot P-value of FST vs P-value of kruskal wallis for population differences in CpG
p2 <- ggplot(dat_bins, aes(x = chromStart, y=OutFLANK_0.2_bigsnpr_0.8.2log10_pvaluesRightTail_unrelated)) + geom_bin2d(binwidth = c(10000, 0.2), color="blue", alpha=0.5) + theme_classic() +
  geom_point(data=dat_points, aes(x=chromStart, y=OutFLANK_0.2_bigsnpr_0.8.2log10_pvaluesRightTail_unrelated),
                                  alpha = 0.5, colour="blue", size=0.3)
 # + add y-axis
p2



p2 + geom_bin2d(data=dat_bins, aes(x = chromStart, y=LFMM_ridge_0.0_log10p_Mean_Annual_Temperature_Celsius_wild_for_assoc), binwidth = c(10000, 0.2), color="blue", alpha=0.5)

Spearmanns_Rho_log10p_Mean_Annual_Temperature_Celsius_wild_for_assoc


cvdat2_chr <- cvdat2 %>% filter(seqid==mychrom)
#p + geom_bin2d(data=cvdat2_chr, aes(x=start, y=meanCpGOE), binwidth=c(10000, 0.01), color="red")
#p + geom_point(data=cvdat2_chr, aes(x=start, y=meanCpGOE), colour="red", alpha=0.7,
#               size=0.3)


p2 + geom_point(data=cvdat2_chr, aes(x=start, y=-log10(kruskal_p)), colour="red", alpha=0.7,
               size=0.5)

head(cvdat2)
cvdat2$numFSToutliers <- NA
cvdat2$Fst_mean <- NA

for (i in 1:nrow(cvdat2)){
  (ind <- which(datchr$chromEnd > cvdat2_chr$start[i] & datchr$chromEnd < cvdat2_chr$end[i]))
  if (length(ind)>0){
    cvdat2$numFSToutliers[i] <- sum(datchr$OutFLANK_0.2_bigsnpr_0.8.2_OutlierFlag_unrelated[ind], na.rm=TRUE)
    cvdat2$Fst_mean[i] <- mean(datchr$OutFLANK_0.2_bigsnpr_0.8.2_FST_unrelated[ind])
    # note this is mean of ratios, need to go back and take mean of num. and denom. then divide
  }
  if(i%%1000==0){print(c(i,"of",nrow(cvdat2)))}
}

head(cvdat2)
hist(cvdat2$Fst_mean)
cvdat2$FST_meanNoNA <- cvdat2$Fst_mean
cvdat2$FST_meanNoNA[is.na(cvdat2$FST_meanNoNA)] <- 0

head(cvdat2)
cvdat2$log10kruskal_p <- -log10(cvdat2$kruskal_p)
cvdat2$numFSToutliers[is.na(cvdat2$numFSToutliers)] <- 0

cvdat_plot <- cvdat2 %>% select(Fst_mean, FST_meanNoNA, numFSToutliers, meanCpGOE, log10kruskal_p, cv_gene)
head(cvdat_plot)
cor(cvdat_plot, method="spearman",use = "na.or.complete")

panel.lm <- function (x, y,  pch = par("pch"), col.lm = "red",  ...) {   
  ymin <- min(y)
  ymax <- max(y)
  xmin <- min(x)
  xmax <- max(x)
  ylim <- c(min(ymin,xmin),max(ymax,xmax))
  xlim <- ylim
  points(x, y, pch = 19,col=adjustcolor("black", 0.1) ,ylim = ylim, xlim= xlim,...)
  ok <- is.finite(x) & is.finite(y)
  if (any(ok))
    abline(lm(y[ok]~ x[ok]),
           col = col.lm, ...)
}

pairs(cvdat_plot, panel=panel.lm)

pairs(cvdat_plot, pch = 19,col=adjustcolor("black", 0.1), panel=panel.smooth)


boxplot(meanCpGOE~is.na(Fst_mean), data= cvdat2)

plot(meanCpGOE ~ Fst_mean, data= cvdat2, col=adjustcolor("black", 0.5))
plot(numFSToutliers ~ Fst_mean, data= cvdat2, col=adjustcolor("black", 0.5))
plot(numFSToutliers/(end-start) ~ Fst_mean, data= cvdat2, col=adjustcolor("black", 0.5), pch=19)
plot(numFSToutliers/(end-start) ~ meanCpGOE, data= cvdat2, col=adjustcolor("black", 0.5), pch=19)
abline(lm(numFSToutliers/(end-start) ~ meanCpGOE, data=cvdat2))

plot(Fst_mean ~ meanCpGOE, data=cvdat2)
plot(FST_meanNoNA ~ meanCpGOE, data=cvdat2)
plot(cv_gene ~ meanCpGOE, data= cvdat2, col=adjustcolor("black", 0.1), pch=19)

plot(cv vs mean cpg0e)
plot(mean Cpg0e)

hist(cvdat2$meanCpGOE, breaks=seq(0,2.5, by=0.01))
hist(cvdat2$meanCpGOE[which(is.na(cvdat2$Fst_mean))], breaks=seq(0,2.5, by=0.01))
